In [1]:
TEST = False
name = 'MRL_ellipse_l'

In [2]:
import logging
import os
import sys
import time

import numpy as np
import matplotlib.pyplot as plt
import pykeen
from pykeen.kge_models import Region

import json

In [3]:
%matplotlib inline

In [4]:
logging.basicConfig(level=logging.INFO)
logging.getLogger('pykeen').setLevel(logging.INFO)

In [5]:
print(sys.version)

3.6.8 |Anaconda, Inc.| (default, Feb 21 2019, 18:30:04) [MSC v.1916 64 bit (AMD64)]


In [ ]:
print(time.asctime())

Fri Jul  5 14:52:42 2019


In [ ]:
print(pykeen.get_version())

0.0.26-dev


Check which hyper-parameters are required by Region:

In [ ]:
Region.hyper_params

['embedding_dim',
 'margin_loss',
 'learning_rate',
 'corrupt_relations',
 'normalization_of_entities',
 'init_radius',
 'reg_lambda',
 'loss_type',
 'neg_factor',
 'region_type']

Define output directory:

In [ ]:
today = time.localtime()
output_directory = 'D:/uni/master/models/region_hpo/opt_{}_{}_{}_{}'.format(today.tm_year, today.tm_mon, today.tm_mday, name)
output_directory

'D:/uni/master/models/region_hpo/opt_2019_7_5_MRL_ellipse_lessdim'

Define hyper-parameters inclduing the evaluation information:
 * Define the hyper-parameter optimization mode: **HPO_mode**
 * Define for each hyper-parameter a list of values (single valued lists are possible)
 * Define the number of hyper-parameter optimization iterations: **maximum_number_of_hpo_iters**
 * Define the path to the test set (**test_set_path**) or define the ratio of the training set that should be used as test set (**test_set_ratio**)
 * Define whether you want to compute the metrics (mean rank and hits@k) in raw or in a filtered setting: **filter_negative_triples**

In [ ]:
config = dict(
    metrics                     = ['mean_rank'],
    training_set_path           = 'D:/uni/master/data/FB15K/train.tsv', 
    test_set_path               = 'D:/uni/master/data/FB15K/valid_150.tsv',
    mapping_path                = 'D:/uni/master/data/fb15k/',
    execution_mode              = 'HPO_mode', # Define hyper-parameter optimization mode
    random_seed                 = 5,
    
    kg_embedding_model_name     = 'Region',
    region_type                 = ['ellipse'],
    embedding_dim               = [30],
    normalization_of_entities   = [2],
    init_radius                 = 0,
    reg_lambda                  = [0.01], # [5, 1, 0.1, 0.01, 0],
    loss_type                   = ['MRL'],
    margin_loss                 = [0.5],  # [0.0, 0.02, 0.05, 0.1, 0.2, 0.3],
    
    neg_factor                  = [1],
    corrupt_relations           = [False], # [True, False],
    filter_negative_triples     = False,
    
    es_metric                   = 'custom',
    learning_rate               = [1e-1],
    num_epochs                  = [500] if not TEST else [10],  
    batch_size                  = [2048],
    test_batch_size             = [4096],
    maximum_number_of_hpo_iters = 1, # Define the number HPO iterations
    preferred_device            = 'gpu'
)

config = dict(
    metrics                     = ['mean_rank'],
    training_set_path           = 'D:/uni/master/data/FB15K/train.tsv', 
    test_set_path               = 'D:/uni/master/data/FB15K/valid_150.tsv',
    mapping_path                = 'D:/uni/master/data/fb15k/',
    execution_mode              = 'HPO_mode', # Define hyper-parameter optimization mode
    random_seed                 = 5,
    
    kg_embedding_model_name     = 'Region',
    region_type                 = ['sphere'],
    embedding_dim               = [50],
    normalization_of_entities   = [1,2],
    init_radius                 = 0,
    reg_lambda                  = [100, 10, 1, 0, 0.1, 0.01],
    loss_type                   = ['NLL'],
    margin_loss                 = [0],  # [0.0, 0.02, 0.05, 0.1, 0.2, 0.3],
    
    neg_factor                  = [1, 2, 5],
    corrupt_relations           = [True, False],
    filter_negative_triples     = False,
    
    es_metric                   = 'custom',
    learning_rate               = [1e-1],
    num_epochs                  = [1000] if not TEST else [10],  
    batch_size                  = [2048],
    test_batch_size             = [4096],
    maximum_number_of_hpo_iters = 10, # Define the number HPO iterations
    preferred_device            = 'gpu'
)

Train and evaluate Region:

In [ ]:
results = pykeen.run(
    config=config,
    output_directory=output_directory
)

HPO Iteration:   0%|                                                                             | 0/1 [00:00<?, ?it/s]

{'metrics': 'mean_rank', 'training_set_path': 'D:/uni/master/data/FB15K/train.tsv', 'test_set_path': 'D:/uni/master/data/FB15K/valid_150.tsv', 'mapping_path': 'D:/uni/master/data/fb15k/', 'execution_mode': 'HPO_mode', 'random_seed': 5, 'kg_embedding_model_name': 'Region', 'region_type': 'ellipse', 'embedding_dim': 30, 'normalization_of_entities': 2, 'init_radius': 0, 'reg_lambda': 0.01, 'loss_type': 'MRL', 'margin_loss': 0.5, 'neg_factor': 1, 'corrupt_relations': False, 'filter_negative_triples': False, 'es_metric': 'custom', 'learning_rate': 0.1, 'num_epochs': 500, 'batch_size': 2048, 'test_batch_size': 4096, 'maximum_number_of_hpo_iters': 1, 'preferred_device': 'gpu', 'pykeen-version': '0.0.26-dev', 'num_entities': 14951, 'num_relations': 1345}


INFO:pykeen.utilities.evaluation_utils.metrics_computations:Evaluation took 15.98s seconds


0.12946576553904768 (0.10006450327967042, 0.18333333333333332)


INFO:pykeen.utilities.train_utils:Epoch   0 /  500 (11.0s):  loss: 0.213  val loss: 0.093  value: 0.129  patience:   1
INFO:pykeen.utilities.train_utils:Epoch   1 /  500 (9.6s):  loss: 0.069  val loss: 0.065  value: 0.129  patience:   2
INFO:pykeen.utilities.train_utils:Epoch   2 /  500 (9.5s):  loss: 0.051  val loss: 0.057  value: 0.129  patience:   3
INFO:pykeen.utilities.train_utils:Epoch   3 /  500 (9.5s):  loss: 0.044  val loss: 0.045  value: 0.129  patience:   4
INFO:pykeen.utilities.train_utils:Epoch   4 /  500 (9.6s):  loss: 0.041  val loss: 0.039  value: 0.129  patience:   5
INFO:pykeen.utilities.train_utils:Epoch   5 /  500 (9.7s):  loss: 0.039  val loss: 0.050  value: 0.129  patience:   6
INFO:pykeen.utilities.train_utils:Epoch   6 /  500 (10.0s):  loss: 0.037  val loss: 0.046  value: 0.129  patience:   7
INFO:pykeen.utilities.train_utils:Epoch   7 /  500 (10.6s):  loss: 0.035  val loss: 0.048  value: 0.129  patience:   8
INFO:pykeen.utilities.train_utils:Epoch   8 /  500 (1

0.15845967060930252 (0.12490439532027958, 0.21666666666666667)


INFO:pykeen.utilities.train_utils:Epoch  10 /  500 (10.2s):  loss: 0.033  val loss: 0.044  value: 0.158  patience:   1
INFO:pykeen.utilities.train_utils:Epoch  11 /  500 (10.5s):  loss: 0.033  val loss: 0.042  value: 0.158  patience:   2
INFO:pykeen.utilities.train_utils:Epoch  12 /  500 (10.5s):  loss: 0.032  val loss: 0.043  value: 0.158  patience:   3
INFO:pykeen.utilities.train_utils:Epoch  13 /  500 (10.1s):  loss: 0.032  val loss: 0.038  value: 0.158  patience:   4
INFO:pykeen.utilities.train_utils:Epoch  14 /  500 (10.2s):  loss: 0.032  val loss: 0.042  value: 0.158  patience:   5
INFO:pykeen.utilities.train_utils:Epoch  15 /  500 (10.1s):  loss: 0.032  val loss: 0.038  value: 0.158  patience:   6
INFO:pykeen.utilities.train_utils:Epoch  16 /  500 (10.1s):  loss: 0.031  val loss: 0.036  value: 0.158  patience:   7
INFO:pykeen.utilities.train_utils:Epoch  17 /  500 (10.5s):  loss: 0.031  val loss: 0.042  value: 0.158  patience:   8
INFO:pykeen.utilities.train_utils:Epoch  18 /  5

0.196425361523384 (0.1532931668194514, 0.2733333333333333)


INFO:pykeen.utilities.train_utils:Epoch  20 /  500 (10.1s):  loss: 0.031  val loss: 0.038  value: 0.196  patience:   1
INFO:pykeen.utilities.train_utils:Epoch  21 /  500 (10.0s):  loss: 0.031  val loss: 0.035  value: 0.196  patience:   2
INFO:pykeen.utilities.train_utils:Epoch  22 /  500 (10.0s):  loss: 0.030  val loss: 0.036  value: 0.196  patience:   3
INFO:pykeen.utilities.train_utils:Epoch  23 /  500 (10.0s):  loss: 0.031  val loss: 0.038  value: 0.196  patience:   4
INFO:pykeen.utilities.train_utils:Epoch  24 /  500 (10.0s):  loss: 0.030  val loss: 0.042  value: 0.196  patience:   5
INFO:pykeen.utilities.train_utils:Epoch  25 /  500 (10.2s):  loss: 0.030  val loss: 0.036  value: 0.196  patience:   6
INFO:pykeen.utilities.train_utils:Epoch  26 /  500 (10.0s):  loss: 0.031  val loss: 0.044  value: 0.196  patience:   7
INFO:pykeen.utilities.train_utils:Epoch  27 /  500 (10.0s):  loss: 0.030  val loss: 0.032  value: 0.196  patience:   8
INFO:pykeen.utilities.train_utils:Epoch  28 /  5

0.1721357109664145 (0.13035479288768295, 0.25333333333333335)


INFO:pykeen.utilities.train_utils:Epoch  30 /  500 (10.0s):  loss: 0.030  val loss: 0.047  value: 0.172  patience:  11
INFO:pykeen.utilities.train_utils:Epoch  31 /  500 (10.1s):  loss: 0.030  val loss: 0.031  value: 0.172  patience:  12
INFO:pykeen.utilities.train_utils:Epoch  32 /  500 (10.0s):  loss: 0.030  val loss: 0.037  value: 0.172  patience:  13
INFO:pykeen.utilities.train_utils:Epoch  33 /  500 (10.0s):  loss: 0.030  val loss: 0.033  value: 0.172  patience:  14
INFO:pykeen.utilities.train_utils:Epoch  34 /  500 (10.0s):  loss: 0.030  val loss: 0.033  value: 0.172  patience:  15
INFO:pykeen.utilities.train_utils:Epoch  35 /  500 (10.0s):  loss: 0.030  val loss: 0.033  value: 0.172  patience:  16
INFO:pykeen.utilities.train_utils:Epoch  36 /  500 (10.1s):  loss: 0.030  val loss: 0.033  value: 0.172  patience:  17
INFO:pykeen.utilities.train_utils:Epoch  37 /  500 (10.1s):  loss: 0.030  val loss: 0.032  value: 0.172  patience:  18
INFO:pykeen.utilities.train_utils:Epoch  38 /  5

0.1818898760686401 (0.1464343769369097, 0.24)


INFO:pykeen.utilities.train_utils:Epoch  40 /  500 (10.0s):  loss: 0.030  val loss: 0.046  value: 0.182  patience:  21
INFO:pykeen.utilities.train_utils:Epoch  41 /  500 (10.0s):  loss: 0.030  val loss: 0.038  value: 0.182  patience:  22
INFO:pykeen.utilities.train_utils:Epoch  42 /  500 (10.0s):  loss: 0.030  val loss: 0.035  value: 0.182  patience:  23
INFO:pykeen.utilities.train_utils:Epoch  43 /  500 (10.0s):  loss: 0.029  val loss: 0.036  value: 0.182  patience:  24
INFO:pykeen.utilities.train_utils:Epoch  44 /  500 (10.0s):  loss: 0.030  val loss: 0.046  value: 0.182  patience:  25
INFO:pykeen.utilities.train_utils:Epoch  45 /  500 (9.9s):  loss: 0.029  val loss: 0.042  value: 0.182  patience:  26
INFO:pykeen.utilities.train_utils:Epoch  46 /  500 (10.0s):  loss: 0.030  val loss: 0.040  value: 0.182  patience:  27
INFO:pykeen.utilities.train_utils:Epoch  47 /  500 (10.0s):  loss: 0.030  val loss: 0.051  value: 0.182  patience:  28
INFO:pykeen.utilities.train_utils:Epoch  48 /  50

0.2064736790076807 (0.16353133186766775, 0.28)


INFO:pykeen.utilities.train_utils:Epoch  50 /  500 (10.0s):  loss: 0.029  val loss: 0.030  value: 0.206  patience:   1
INFO:pykeen.utilities.train_utils:Epoch  51 /  500 (10.0s):  loss: 0.030  val loss: 0.033  value: 0.206  patience:   2
INFO:pykeen.utilities.train_utils:Epoch  52 /  500 (10.0s):  loss: 0.029  val loss: 0.032  value: 0.206  patience:   3
INFO:pykeen.utilities.train_utils:Epoch  53 /  500 (10.1s):  loss: 0.029  val loss: 0.046  value: 0.206  patience:   4
INFO:pykeen.utilities.train_utils:Epoch  54 /  500 (10.0s):  loss: 0.029  val loss: 0.040  value: 0.206  patience:   5
INFO:pykeen.utilities.train_utils:Epoch  55 /  500 (10.0s):  loss: 0.030  val loss: 0.035  value: 0.206  patience:   6
INFO:pykeen.utilities.train_utils:Epoch  56 /  500 (10.1s):  loss: 0.030  val loss: 0.034  value: 0.206  patience:   7
INFO:pykeen.utilities.train_utils:Epoch  57 /  500 (10.0s):  loss: 0.030  val loss: 0.039  value: 0.206  patience:   8
INFO:pykeen.utilities.train_utils:Epoch  58 /  5

0.18039905679854995 (0.13719182390476775, 0.2633333333333333)


INFO:pykeen.utilities.train_utils:Epoch  60 /  500 (10.0s):  loss: 0.029  val loss: 0.042  value: 0.180  patience:  11
INFO:pykeen.utilities.train_utils:Epoch  61 /  500 (10.2s):  loss: 0.029  val loss: 0.035  value: 0.180  patience:  12
INFO:pykeen.utilities.evaluation_utils.metrics_computations:Evaluation took 17.49s seconds


0.17712888407849134 (0.13260465776440783, 0.26666666666666666)


INFO:pykeen.utilities.train_utils:Epoch  62 /  500 (10.1s):  loss: 0.029  val loss: 0.029  value: 0.177  patience:  13
INFO:pykeen.utilities.evaluation_utils.metrics_computations:Evaluation took 17.46s seconds


0.20054530732188056 (0.15421540058757027, 0.2866666666666667)


INFO:pykeen.utilities.train_utils:Epoch  63 /  500 (9.9s):  loss: 0.029  val loss: 0.028  value: 0.201  patience:  14
INFO:pykeen.utilities.train_utils:Epoch  64 /  500 (10.0s):  loss: 0.029  val loss: 0.029  value: 0.201  patience:  15
INFO:pykeen.utilities.train_utils:Epoch  65 /  500 (10.0s):  loss: 0.029  val loss: 0.041  value: 0.201  patience:  16
INFO:pykeen.utilities.train_utils:Epoch  66 /  500 (10.1s):  loss: 0.029  val loss: 0.041  value: 0.201  patience:  17
INFO:pykeen.utilities.train_utils:Epoch  67 /  500 (10.5s):  loss: 0.029  val loss: 0.036  value: 0.201  patience:  18
INFO:pykeen.utilities.train_utils:Epoch  68 /  500 (10.3s):  loss: 0.029  val loss: 0.031  value: 0.201  patience:  19
INFO:pykeen.utilities.train_utils:Epoch  69 /  500 (10.2s):  loss: 0.029  val loss: 0.033  value: 0.201  patience:  20
INFO:pykeen.utilities.evaluation_utils.metrics_computations:Evaluation took 17.36s seconds


0.17265553931550787 (0.13280734481454176, 0.24666666666666667)


INFO:pykeen.utilities.train_utils:Epoch  70 /  500 (10.1s):  loss: 0.029  val loss: 0.045  value: 0.173  patience:  21
INFO:pykeen.utilities.train_utils:Epoch  71 /  500 (10.1s):  loss: 0.029  val loss: 0.037  value: 0.173  patience:  22
INFO:pykeen.utilities.train_utils:Epoch  72 /  500 (10.1s):  loss: 0.029  val loss: 0.037  value: 0.173  patience:  23
INFO:pykeen.utilities.train_utils:Epoch  73 /  500 (10.1s):  loss: 0.029  val loss: 0.037  value: 0.173  patience:  24
INFO:pykeen.utilities.train_utils:Epoch  74 /  500 (10.2s):  loss: 0.029  val loss: 0.038  value: 0.173  patience:  25
INFO:pykeen.utilities.train_utils:Epoch  75 /  500 (10.1s):  loss: 0.029  val loss: 0.035  value: 0.173  patience:  26
INFO:pykeen.utilities.train_utils:Epoch  76 /  500 (9.9s):  loss: 0.029  val loss: 0.041  value: 0.173  patience:  27
INFO:pykeen.utilities.train_utils:Epoch  77 /  500 (9.9s):  loss: 0.029  val loss: 0.036  value: 0.173  patience:  28
INFO:pykeen.utilities.train_utils:Epoch  78 /  500

0.18318732293437998 (0.14688176587515844, 0.24333333333333335)


INFO:pykeen.utilities.train_utils:Epoch  80 /  500 (9.9s):  loss: 0.029  val loss: 0.040  value: 0.183  patience:  31
INFO:pykeen.utilities.train_utils:Epoch  81 /  500 (10.0s):  loss: 0.029  val loss: 0.030  value: 0.183  patience:  32
INFO:pykeen.utilities.train_utils:Epoch  82 /  500 (10.0s):  loss: 0.029  val loss: 0.034  value: 0.183  patience:  33
INFO:pykeen.utilities.train_utils:Epoch  83 /  500 (9.9s):  loss: 0.029  val loss: 0.032  value: 0.183  patience:  34


Epoch    84: reducing learning rate of group 0 to 1.0000e-02.


INFO:pykeen.utilities.train_utils:Epoch  84 /  500 (10.9s):  loss: 0.029  val loss: 0.036  value: 0.183  patience:  35
INFO:pykeen.utilities.train_utils:Epoch  85 /  500 (10.0s):  loss: 0.023  val loss: 0.034  value: 0.183  patience:  36
INFO:pykeen.utilities.train_utils:Epoch  86 /  500 (9.9s):  loss: 0.022  val loss: 0.032  value: 0.183  patience:  37
INFO:pykeen.utilities.train_utils:Epoch  87 /  500 (9.9s):  loss: 0.021  val loss: 0.035  value: 0.183  patience:  38
INFO:pykeen.utilities.train_utils:Epoch  88 /  500 (9.9s):  loss: 0.021  val loss: 0.030  value: 0.183  patience:  39
INFO:pykeen.utilities.train_utils:Epoch  89 /  500 (9.9s):  loss: 0.021  val loss: 0.031  value: 0.183  patience:  40
INFO:pykeen.utilities.evaluation_utils.metrics_computations:Evaluation took 17.60s seconds


0.21180422122220754 (0.17032136963402073, 0.28)


INFO:pykeen.utilities.train_utils:Epoch  90 /  500 (10.0s):  loss: 0.020  val loss: 0.030  value: 0.212  patience:   1
INFO:pykeen.utilities.train_utils:Epoch  91 /  500 (9.9s):  loss: 0.020  val loss: 0.029  value: 0.212  patience:   2
INFO:pykeen.utilities.train_utils:Epoch  92 /  500 (10.0s):  loss: 0.020  val loss: 0.031  value: 0.212  patience:   3
INFO:pykeen.utilities.train_utils:Epoch  93 /  500 (9.9s):  loss: 0.020  val loss: 0.031  value: 0.212  patience:   4
INFO:pykeen.utilities.train_utils:Epoch  94 /  500 (10.0s):  loss: 0.020  val loss: 0.031  value: 0.212  patience:   5
INFO:pykeen.utilities.train_utils:Epoch  95 /  500 (9.9s):  loss: 0.019  val loss: 0.029  value: 0.212  patience:   6
INFO:pykeen.utilities.train_utils:Epoch  96 /  500 (9.9s):  loss: 0.019  val loss: 0.029  value: 0.212  patience:   7
INFO:pykeen.utilities.train_utils:Epoch  97 /  500 (9.9s):  loss: 0.019  val loss: 0.028  value: 0.212  patience:   8
INFO:pykeen.utilities.train_utils:Epoch  98 /  500 (9

0.22267304349571362 (0.17703986396199933, 0.3)


INFO:pykeen.utilities.train_utils:Epoch 100 /  500 (10.0s):  loss: 0.019  val loss: 0.027  value: 0.223  patience:   1
INFO:pykeen.utilities.train_utils:Epoch 101 /  500 (10.0s):  loss: 0.019  val loss: 0.028  value: 0.223  patience:   2
INFO:pykeen.utilities.evaluation_utils.metrics_computations:Evaluation took 17.25s seconds


0.22469540876536276 (0.17620735569933627, 0.31)


INFO:pykeen.utilities.train_utils:Epoch 102 /  500 (10.0s):  loss: 0.019  val loss: 0.027  value: 0.225  patience:   1
INFO:pykeen.utilities.evaluation_utils.metrics_computations:Evaluation took 17.61s seconds


0.22136936509178043 (0.17427764777148955, 0.30333333333333334)


INFO:pykeen.utilities.train_utils:Epoch 103 /  500 (10.1s):  loss: 0.019  val loss: 0.026  value: 0.221  patience:   2
INFO:pykeen.utilities.evaluation_utils.metrics_computations:Evaluation took 18.20s seconds


0.23023165876363014 (0.17775634805969628, 0.32666666666666666)


INFO:pykeen.utilities.train_utils:Epoch 104 /  500 (11.1s):  loss: 0.018  val loss: 0.025  value: 0.230  patience:   1
INFO:pykeen.utilities.train_utils:Epoch 105 /  500 (10.0s):  loss: 0.018  val loss: 0.025  value: 0.230  patience:   2
INFO:pykeen.utilities.evaluation_utils.metrics_computations:Evaluation took 17.14s seconds


0.22679269235610613 (0.17992871024732265, 0.30666666666666664)


INFO:pykeen.utilities.train_utils:Epoch 106 /  500 (10.3s):  loss: 0.018  val loss: 0.024  value: 0.227  patience:   3
INFO:pykeen.utilities.train_utils:Epoch 107 /  500 (9.9s):  loss: 0.018  val loss: 0.027  value: 0.227  patience:   4
INFO:pykeen.utilities.train_utils:Epoch 108 /  500 (9.9s):  loss: 0.018  val loss: 0.027  value: 0.227  patience:   5
INFO:pykeen.utilities.train_utils:Epoch 109 /  500 (9.9s):  loss: 0.018  val loss: 0.027  value: 0.227  patience:   6
INFO:pykeen.utilities.evaluation_utils.metrics_computations:Evaluation took 17.48s seconds


0.23083726006974842 (0.17949038613383742, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 110 /  500 (9.9s):  loss: 0.018  val loss: 0.024  value: 0.231  patience:   1
INFO:pykeen.utilities.train_utils:Epoch 111 /  500 (9.9s):  loss: 0.018  val loss: 0.025  value: 0.231  patience:   2
INFO:pykeen.utilities.train_utils:Epoch 112 /  500 (9.9s):  loss: 0.018  val loss: 0.026  value: 0.231  patience:   3
INFO:pykeen.utilities.train_utils:Epoch 113 /  500 (9.9s):  loss: 0.018  val loss: 0.027  value: 0.231  patience:   4
INFO:pykeen.utilities.train_utils:Epoch 114 /  500 (9.8s):  loss: 0.018  val loss: 0.025  value: 0.231  patience:   5
INFO:pykeen.utilities.train_utils:Epoch 115 /  500 (9.9s):  loss: 0.018  val loss: 0.027  value: 0.231  patience:   6
INFO:pykeen.utilities.train_utils:Epoch 116 /  500 (9.9s):  loss: 0.018  val loss: 0.027  value: 0.231  patience:   7
INFO:pykeen.utilities.train_utils:Epoch 117 /  500 (9.9s):  loss: 0.018  val loss: 0.029  value: 0.231  patience:   8
INFO:pykeen.utilities.train_utils:Epoch 118 /  500 (9.9s

0.22748206218038236 (0.1785582710742264, 0.31333333333333335)


INFO:pykeen.utilities.train_utils:Epoch 120 /  500 (10.0s):  loss: 0.017  val loss: 0.026  value: 0.227  patience:  11
INFO:pykeen.utilities.train_utils:Epoch 121 /  500 (9.9s):  loss: 0.018  val loss: 0.026  value: 0.227  patience:  12
INFO:pykeen.utilities.train_utils:Epoch 122 /  500 (10.0s):  loss: 0.017  val loss: 0.027  value: 0.227  patience:  13
INFO:pykeen.utilities.train_utils:Epoch 123 /  500 (9.9s):  loss: 0.018  val loss: 0.025  value: 0.227  patience:  14
INFO:pykeen.utilities.train_utils:Epoch 124 /  500 (10.0s):  loss: 0.017  val loss: 0.024  value: 0.227  patience:  15
INFO:pykeen.utilities.train_utils:Epoch 125 /  500 (9.9s):  loss: 0.017  val loss: 0.026  value: 0.227  patience:  16
INFO:pykeen.utilities.train_utils:Epoch 126 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.227  patience:  17


Epoch   127: reducing learning rate of group 0 to 1.0000e-03.


INFO:pykeen.utilities.train_utils:Epoch 127 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.227  patience:  18
INFO:pykeen.utilities.train_utils:Epoch 128 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.227  patience:  19
INFO:pykeen.utilities.train_utils:Epoch 129 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.227  patience:  20
INFO:pykeen.utilities.evaluation_utils.metrics_computations:Evaluation took 17.47s seconds


0.2347250978586152 (0.18886457418258748, 0.31)


INFO:pykeen.utilities.train_utils:Epoch 130 /  500 (10.0s):  loss: 0.017  val loss: 0.026  value: 0.235  patience:   1
INFO:pykeen.utilities.train_utils:Epoch 131 /  500 (10.0s):  loss: 0.017  val loss: 0.026  value: 0.235  patience:   2
INFO:pykeen.utilities.train_utils:Epoch 132 /  500 (9.9s):  loss: 0.017  val loss: 0.026  value: 0.235  patience:   3
INFO:pykeen.utilities.train_utils:Epoch 133 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.235  patience:   4
INFO:pykeen.utilities.train_utils:Epoch 134 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.235  patience:   5
INFO:pykeen.utilities.train_utils:Epoch 135 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.235  patience:   6
INFO:pykeen.utilities.train_utils:Epoch 136 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.235  patience:   7
INFO:pykeen.utilities.train_utils:Epoch 137 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.235  patience:   8
INFO:pykeen.utilities.train_utils:Epoch 138 /  500 (9

0.23736891434325683 (0.18983226590682245, 0.31666666666666665)


INFO:pykeen.utilities.train_utils:Epoch 140 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:   1
INFO:pykeen.utilities.train_utils:Epoch 141 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:   2
INFO:pykeen.utilities.train_utils:Epoch 142 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:   3
INFO:pykeen.utilities.train_utils:Epoch 143 /  500 (10.0s):  loss: 0.017  val loss: 0.024  value: 0.237  patience:   4
INFO:pykeen.utilities.train_utils:Epoch 144 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:   5
INFO:pykeen.utilities.train_utils:Epoch 145 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:   6
INFO:pykeen.utilities.train_utils:Epoch 146 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:   7
INFO:pykeen.utilities.train_utils:Epoch 147 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:   8


Epoch   148: reducing learning rate of group 0 to 1.0000e-04.


INFO:pykeen.utilities.train_utils:Epoch 148 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:   9
INFO:pykeen.utilities.train_utils:Epoch 149 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  10
INFO:pykeen.utilities.evaluation_utils.metrics_computations:Evaluation took 17.30s seconds


0.23679960278526432 (0.189105075067278, 0.31666666666666665)


INFO:pykeen.utilities.train_utils:Epoch 150 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  11
INFO:pykeen.utilities.train_utils:Epoch 151 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  12
INFO:pykeen.utilities.train_utils:Epoch 152 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  13
INFO:pykeen.utilities.train_utils:Epoch 153 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  14
INFO:pykeen.utilities.train_utils:Epoch 154 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  15
INFO:pykeen.utilities.train_utils:Epoch 155 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  16
INFO:pykeen.utilities.train_utils:Epoch 156 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  17
INFO:pykeen.utilities.train_utils:Epoch 157 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  18
INFO:pykeen.utilities.train_utils:Epoch 158 /  500 (9.9

0.23854038660434468 (0.1889808672051874, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 160 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   1
INFO:pykeen.utilities.train_utils:Epoch 161 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   2
INFO:pykeen.utilities.train_utils:Epoch 162 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   3
INFO:pykeen.utilities.train_utils:Epoch 163 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   4
INFO:pykeen.utilities.train_utils:Epoch 164 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   5
INFO:pykeen.utilities.train_utils:Epoch 165 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   6
INFO:pykeen.utilities.train_utils:Epoch 166 /  500 (10.4s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   7
INFO:pykeen.utilities.train_utils:Epoch 167 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   8
INFO:pykeen.utilities.train_utils:Epoch 168 /  50

0.2369975464952817 (0.18705145915826724, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 170 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  11
INFO:pykeen.utilities.train_utils:Epoch 171 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  12
INFO:pykeen.utilities.train_utils:Epoch 172 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  13
INFO:pykeen.utilities.train_utils:Epoch 173 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  14
INFO:pykeen.utilities.train_utils:Epoch 174 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  15
INFO:pykeen.utilities.train_utils:Epoch 175 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  16
INFO:pykeen.utilities.train_utils:Epoch 176 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  17
INFO:pykeen.utilities.train_utils:Epoch 177 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  18
INFO:pykeen.utilities.train_utils:Epoch 178 /  50

0.2377046834764535 (0.18793396652137986, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 180 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  21
INFO:pykeen.utilities.train_utils:Epoch 181 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  22
INFO:pykeen.utilities.train_utils:Epoch 182 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  23
INFO:pykeen.utilities.train_utils:Epoch 183 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  24
INFO:pykeen.utilities.train_utils:Epoch 184 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  25
INFO:pykeen.utilities.train_utils:Epoch 185 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  26
INFO:pykeen.utilities.train_utils:Epoch 186 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  27
INFO:pykeen.utilities.train_utils:Epoch 187 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  28
INFO:pykeen.utilities.train_utils:Epoch 188 /  500 (9.

0.2376510426515167 (0.1878669157826863, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 190 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  31
INFO:pykeen.utilities.train_utils:Epoch 191 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  32
INFO:pykeen.utilities.train_utils:Epoch 192 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  33
INFO:pykeen.utilities.train_utils:Epoch 193 /  500 (10.5s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  34
INFO:pykeen.utilities.train_utils:Epoch 194 /  500 (10.8s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  35
INFO:pykeen.utilities.train_utils:Epoch 195 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  36
INFO:pykeen.utilities.train_utils:Epoch 196 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  37
INFO:pykeen.utilities.train_utils:Epoch 197 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  38
INFO:pykeen.utilities.train_utils:Epoch 198 /  50

0.23928084149409648 (0.18991203744096405, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 200 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   1
INFO:pykeen.utilities.train_utils:Epoch 201 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   2
INFO:pykeen.utilities.train_utils:Epoch 202 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   3
INFO:pykeen.utilities.train_utils:Epoch 203 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   4
INFO:pykeen.utilities.train_utils:Epoch 204 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   5
INFO:pykeen.utilities.train_utils:Epoch 205 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   6
INFO:pykeen.utilities.train_utils:Epoch 206 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   7
INFO:pykeen.utilities.train_utils:Epoch 207 /  500 (10.4s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   8
INFO:pykeen.utilities.train_utils:Epoch 208 /  500 (

0.23906630737318688 (0.1896418987084545, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 210 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  11
INFO:pykeen.utilities.train_utils:Epoch 211 /  500 (9.8s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  12
INFO:pykeen.utilities.train_utils:Epoch 212 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  13
INFO:pykeen.utilities.train_utils:Epoch 213 /  500 (9.8s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  14
INFO:pykeen.utilities.train_utils:Epoch 214 /  500 (9.8s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  15
INFO:pykeen.utilities.train_utils:Epoch 215 /  500 (9.8s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  16
INFO:pykeen.utilities.train_utils:Epoch 216 /  500 (9.8s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  17
INFO:pykeen.utilities.train_utils:Epoch 217 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  18
INFO:pykeen.utilities.train_utils:Epoch 218 /  500 (9.9s

0.23926794245118205 (0.18989578708555124, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 220 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  21
INFO:pykeen.utilities.train_utils:Epoch 221 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  22
INFO:pykeen.utilities.train_utils:Epoch 222 /  500 (9.8s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  23
INFO:pykeen.utilities.train_utils:Epoch 223 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  24
INFO:pykeen.utilities.train_utils:Epoch 224 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  25
INFO:pykeen.utilities.train_utils:Epoch 225 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  26
INFO:pykeen.utilities.train_utils:Epoch 226 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  27
INFO:pykeen.utilities.train_utils:Epoch 227 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  28
INFO:pykeen.utilities.train_utils:Epoch 228 /  500 (10

0.23903169899202498 (0.18959834689429494, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 230 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  31
INFO:pykeen.utilities.train_utils:Epoch 231 /  500 (11.8s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  32
INFO:pykeen.utilities.train_utils:Epoch 232 /  500 (10.8s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  33
INFO:pykeen.utilities.train_utils:Epoch 233 /  500 (11.1s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  34
INFO:pykeen.utilities.train_utils:Epoch 234 /  500 (10.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  35
INFO:pykeen.utilities.train_utils:Epoch 235 /  500 (10.7s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  36
INFO:pykeen.utilities.train_utils:Epoch 236 /  500 (10.5s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  37
INFO:pykeen.utilities.train_utils:Epoch 237 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  38
INFO:pykeen.utilities.train_utils:Epoch 238 /  5

0.2393426833251847 (0.1899899607521962, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 240 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   1
INFO:pykeen.utilities.train_utils:Epoch 241 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   2
INFO:pykeen.utilities.train_utils:Epoch 242 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   3
INFO:pykeen.utilities.train_utils:Epoch 243 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   4
INFO:pykeen.utilities.train_utils:Epoch 244 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   5
INFO:pykeen.utilities.train_utils:Epoch 245 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   6
INFO:pykeen.utilities.train_utils:Epoch 246 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   7
INFO:pykeen.utilities.train_utils:Epoch 247 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:   8
INFO:pykeen.utilities.train_utils:Epoch 248 /  500 (

0.2392347255909323 (0.18985394485805174, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 250 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  11
INFO:pykeen.utilities.train_utils:Epoch 251 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  12
INFO:pykeen.utilities.train_utils:Epoch 252 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  13
INFO:pykeen.utilities.train_utils:Epoch 253 /  500 (10.5s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  14
INFO:pykeen.utilities.train_utils:Epoch 254 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  15
INFO:pykeen.utilities.train_utils:Epoch 255 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  16
INFO:pykeen.utilities.train_utils:Epoch 256 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  17
INFO:pykeen.utilities.train_utils:Epoch 257 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  18
INFO:pykeen.utilities.train_utils:Epoch 258 /  5

0.23742478867184647 (0.1875842929696506, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 260 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  21
INFO:pykeen.utilities.train_utils:Epoch 261 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  22
INFO:pykeen.utilities.train_utils:Epoch 262 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  23
INFO:pykeen.utilities.train_utils:Epoch 263 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  24
INFO:pykeen.utilities.train_utils:Epoch 264 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  25
INFO:pykeen.utilities.train_utils:Epoch 265 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  26
INFO:pykeen.utilities.train_utils:Epoch 266 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  27
INFO:pykeen.utilities.train_utils:Epoch 267 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.237  patience:  28
INFO:pykeen.utilities.train_utils:Epoch 268 /  5

0.23886684360777533 (0.18939099136234921, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 270 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  31
INFO:pykeen.utilities.train_utils:Epoch 271 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  32
INFO:pykeen.utilities.train_utils:Epoch 272 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  33
INFO:pykeen.utilities.train_utils:Epoch 273 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  34
INFO:pykeen.utilities.train_utils:Epoch 274 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  35
INFO:pykeen.utilities.train_utils:Epoch 275 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  36
INFO:pykeen.utilities.train_utils:Epoch 276 /  500 (10.5s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  37
INFO:pykeen.utilities.train_utils:Epoch 277 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  38
INFO:pykeen.utilities.train_utils:Epoch 278 /  5

0.23902584112012285 (0.18959097597389385, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 280 /  500 (10.8s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  41
INFO:pykeen.utilities.train_utils:Epoch 281 /  500 (10.5s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  42
INFO:pykeen.utilities.train_utils:Epoch 282 /  500 (10.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  43
INFO:pykeen.utilities.train_utils:Epoch 283 /  500 (10.4s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  44
INFO:pykeen.utilities.train_utils:Epoch 284 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  45
INFO:pykeen.utilities.train_utils:Epoch 285 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  46
INFO:pykeen.utilities.train_utils:Epoch 286 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  47
INFO:pykeen.utilities.train_utils:Epoch 287 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  48
INFO:pykeen.utilities.train_utils:Epoch 288 /  50

0.23920630137346907 (0.18981814518761217, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 290 /  500 (10.4s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  51
INFO:pykeen.utilities.train_utils:Epoch 291 /  500 (10.5s):  loss: 0.016  val loss: 0.025  value: 0.239  patience:  52
INFO:pykeen.utilities.train_utils:Epoch 292 /  500 (10.5s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  53
INFO:pykeen.utilities.train_utils:Epoch 293 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  54
INFO:pykeen.utilities.train_utils:Epoch 294 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  55
INFO:pykeen.utilities.train_utils:Epoch 295 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  56
INFO:pykeen.utilities.train_utils:Epoch 296 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  57
INFO:pykeen.utilities.train_utils:Epoch 297 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  58
INFO:pykeen.utilities.train_utils:Epoch 298 /  5

0.23811276773791176 (0.18844464611351208, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 300 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  61
INFO:pykeen.utilities.train_utils:Epoch 301 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  62
INFO:pykeen.utilities.train_utils:Epoch 302 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  63
INFO:pykeen.utilities.train_utils:Epoch 303 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  64
INFO:pykeen.utilities.train_utils:Epoch 304 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  65
INFO:pykeen.utilities.train_utils:Epoch 305 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  66
INFO:pykeen.utilities.train_utils:Epoch 306 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  67
INFO:pykeen.utilities.train_utils:Epoch 307 /  500 (10.8s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  68
INFO:pykeen.utilities.train_utils:Epoch 308 /  5

0.23811613560935957 (0.1884488649366773, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 310 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  71
INFO:pykeen.utilities.train_utils:Epoch 311 /  500 (10.0s):  loss: 0.016  val loss: 0.025  value: 0.238  patience:  72
INFO:pykeen.utilities.train_utils:Epoch 312 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  73
INFO:pykeen.utilities.train_utils:Epoch 313 /  500 (9.9s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  74
INFO:pykeen.utilities.train_utils:Epoch 314 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  75
INFO:pykeen.utilities.train_utils:Epoch 315 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  76
INFO:pykeen.utilities.train_utils:Epoch 316 /  500 (10.4s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  77
INFO:pykeen.utilities.train_utils:Epoch 317 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.238  patience:  78
INFO:pykeen.utilities.train_utils:Epoch 318 /  50

0.2396152228277822 (0.190333654104374, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 320 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.240  patience:   1
INFO:pykeen.utilities.train_utils:Epoch 321 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.240  patience:   2
INFO:pykeen.utilities.train_utils:Epoch 322 /  500 (10.7s):  loss: 0.017  val loss: 0.025  value: 0.240  patience:   3
INFO:pykeen.utilities.train_utils:Epoch 323 /  500 (10.6s):  loss: 0.016  val loss: 0.025  value: 0.240  patience:   4
INFO:pykeen.utilities.train_utils:Epoch 324 /  500 (10.8s):  loss: 0.017  val loss: 0.025  value: 0.240  patience:   5
INFO:pykeen.utilities.train_utils:Epoch 325 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.240  patience:   6
INFO:pykeen.utilities.train_utils:Epoch 326 /  500 (10.8s):  loss: 0.017  val loss: 0.025  value: 0.240  patience:   7
INFO:pykeen.utilities.train_utils:Epoch 327 /  500 (11.3s):  loss: 0.017  val loss: 0.025  value: 0.240  patience:   8
INFO:pykeen.utilities.train_utils:Epoch 328 /  5

0.23862225553853209 (0.19024254041484534, 0.32)


INFO:pykeen.utilities.train_utils:Epoch 330 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  11
INFO:pykeen.utilities.train_utils:Epoch 331 /  500 (10.4s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  12
INFO:pykeen.utilities.train_utils:Epoch 332 /  500 (10.7s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  13
INFO:pykeen.utilities.train_utils:Epoch 333 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  14
INFO:pykeen.utilities.train_utils:Epoch 334 /  500 (10.1s):  loss: 0.016  val loss: 0.025  value: 0.239  patience:  15
INFO:pykeen.utilities.train_utils:Epoch 335 /  500 (10.0s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  16
INFO:pykeen.utilities.train_utils:Epoch 336 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  17
INFO:pykeen.utilities.train_utils:Epoch 337 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  18
INFO:pykeen.utilities.train_utils:Epoch 338 /  5

0.23851122919949672 (0.1901014396782807, 0.32)


INFO:pykeen.utilities.train_utils:Epoch 340 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  21
INFO:pykeen.utilities.train_utils:Epoch 341 /  500 (10.1s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  22
INFO:pykeen.utilities.train_utils:Epoch 342 /  500 (10.8s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  23
INFO:pykeen.utilities.train_utils:Epoch 343 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  24
INFO:pykeen.utilities.train_utils:Epoch 344 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  25
INFO:pykeen.utilities.train_utils:Epoch 345 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  26
INFO:pykeen.utilities.train_utils:Epoch 346 /  500 (10.5s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  27
INFO:pykeen.utilities.train_utils:Epoch 347 /  500 (10.1s):  loss: 0.016  val loss: 0.025  value: 0.239  patience:  28
INFO:pykeen.utilities.train_utils:Epoch 348 /  5

0.23936674428217866 (0.190020284965446, 0.3233333333333333)


INFO:pykeen.utilities.train_utils:Epoch 350 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  31
INFO:pykeen.utilities.train_utils:Epoch 351 /  500 (10.3s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  32
INFO:pykeen.utilities.train_utils:Epoch 352 /  500 (10.2s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  33
INFO:pykeen.utilities.train_utils:Epoch 353 /  500 (10.6s):  loss: 0.017  val loss: 0.025  value: 0.239  patience:  34


In [ ]:
config['reg_lambda'] = [0.1]
results30 = pykeen.run(
    config=config,
    output_directory=output_directory
)

In [ ]:
config['reg_lambda'] = [1]
results40 = pykeen.run(
    config=config,
    output_directory=output_directory
)

In [ ]:
config['reg_lambda'] = [5]
results40 = pykeen.run(
    config=config,
    output_directory=output_directory
)

Check result entries:

In [ ]:
results.results.keys()

In [ ]:
results = {
    'losses': json.load(open(os.path.join(output_directory, 'losses.json'))),
    'validation_losses': json.load(open(os.path.join(output_directory, 'val_losses.json'))),
    'entity_to_embedding': json.load(open(os.path.join(output_directory, 'entities_to_embeddings.json'))),
    'relation_to_embedding': json.load(open(os.path.join(output_directory, 'relations_to_embeddings.json'))),
    'final_configuration': json.load(open(os.path.join(output_directory, 'configuration.json'))),
    'eval_summary': json.load(open(os.path.join(output_directory, 'evaluation_summary.json'))),
    'search_summary': json.load(open(os.path.join(output_directory, 'search_summary.json'))),
    'entity_to_id': json.load(open(os.path.join(output_directory, 'entity_to_id.json'))),
    'relation_to_id': json.load(open(os.path.join(output_directory, 'relation_to_id.json')))
}

In [ ]:
results['search_summary']

In [ ]:
import torch, os
import pykeen.constants as pkc
from pykeen.kge_models import get_kge_model
# Initialize KG embedding model
config = results['final_configuration']
if config['random_seed'] is not None:
    torch.manual_seed(config['random_seed'])
    
model: torch.nn.Module = get_kge_model(config=config)
model.load_state_dict(torch.load(os.path.join(output_directory, 'trained_model.pkl')))

results['trained_model'] = model 

Get the model configuration that lead to best results:

In [ ]:
results['final_configuration']

Access trained model:

In [ ]:
results['trained_model']

Visualize loss values:

In [ ]:
losses = results['losses']
epochs = np.arange(len(losses))
plt.title(r'Loss Per Epoch')
plt.xlabel('epoch')
plt.ylabel('loss')
plt.plot(epochs, losses)
plt.show()

In [ ]:
losses = results['validation_losses']
epochs = np.arange(len(losses))
plt.title(r'Loss Per Epoch')
plt.xlabel('epoch')
plt.ylabel('loss')
plt.plot(epochs, losses)
plt.show()

Get evaluation results:
 * Mean rank
 * Hits@k, k $\in$ {1,3,5,10} 

In [ ]:
results['eval_summary']

1.

    {'training_set_path': 'D:/uni/master/data/YAGO39K/Train/triple2id.tsv',
     'test_set_path': 'D:/uni/master/data/YAGO39K/Valid/instanceOf2id_positive.tsv',
     'neg_test_set_path': 'D:/uni/master/data/YAGO39K/Valid/instanceOf2id_negative.tsv',
     'execution_mode': 'HPO_mode',
     'random_seed': 2,
     'kg_embedding_model_name': 'Region',
     'embedding_dim': 20,
     'normalization_of_entities': 2,
     'init_radius': 5,
     'reg_lambda': 0.02,
     'margin_loss': 0,
     'learning_rate': 5,
     'num_epochs': 30,
     'batch_size': 1024,
     'test_batch_size': 1024,
     'filter_negative_triples': True,
     'maximum_number_of_hpo_iters': 20,
     'preferred_device': 'gpu',
     'pykeen-version': '0.0.26-dev',
     'num_entities': 46100,
     'num_relations': 39}


     {'mean_rank': None,
     'hits@k': None,
     'precision': [0.5817929759704251],
     'recall': [0.7386089563913696],
     'accuracy': [0.6035163257983495],
     'f1_score': 0.6508889335669357}

     30 epochs overfit

2.
   
       {'training_set_path': 'D:/uni/master/data/YAGO39K/Train/instanceOf2id.tsv', 'test_set_path': 'D:/uni/master/data/YAGO39K/Valid/subClassOf2id_positive.tsv', 'neg_test_set_path': 'D:/uni/master/data/YAGO39K/Valid/triple2id_negative.tsv', 'execution_mode': 'HPO_mode', 'random_seed': 2, 'kg_embedding_model_name': 'Region', 'embedding_dim': 100, 'normalization_of_entities': 2, 'init_radius': 0.5, 'reg_lambda': 0.2, 'margin_loss': 0, 'learning_rate': 5, 'num_epochs': 30, 'batch_size': 1024, 'test_batch_size': 1024, 'filter_negative_triples': True, 'maximum_number_of_hpo_iters': 20, 'preferred_device': 'gpu', 'pykeen-version': '0.0.26-dev', 'num_entities': 46100, 'num_relations': 39}

    MetricResults(mean_rank=None, hits_at_k=None, precision=(0.5531790499390986,), recall=(0.7401082067661822,), accuracy=(0.5707994911439476,), fscore=0.6331344448781575)

    18 epochs


3.

    {'training_set_path': 'D:/uni/master/data/YAGO39K/Train/instanceOf2id.tsv', 'test_set_path': 'D:/uni/master/data/YAGO39K/Valid/instanceOf2id_positive.tsv', 'neg_test_set_path': 'D:/uni/master/data/YAGO39K/Valid/triple2id_negative.tsv', 'execution_mode': 'HPO_mode', 'random_seed': 2, 'kg_embedding_model_name': 'Region', 'embedding_dim': 50, 'normalization_of_entities': 2, 'init_radius': 10, 'reg_lambda': 0.1, 'margin_loss': 0, 'learning_rate': 2, 'num_epochs': 30, 'batch_size': 1024, 'test_batch_size': 1024, 'filter_negative_triples': True, 'maximum_number_of_hpo_iters': 20, 'preferred_device': 'gpu', 'pykeen-version': '0.0.26-dev', 'num_entities': 46100, 'num_relations': 39}

    MetricResults(mean_rank=None, hits_at_k=None, precision=(0.5531279471864193,), recall=(0.6881559220389805,), accuracy=(0.5657435495971557,), fscore=0.6132977023847561)

    30 epochs not finished



        {'metrics': 'mean_rank', 'training_set_path': 'D:/uni/master/data/FB15K/train.tsv', 'test_set_path': 'D:/uni/master/data/FB15K/valid_150.tsv', 'mapping_path': 'D:/uni/master/data/fb15k/', 'execution_mode': 'HPO_mode', 'random_seed': 5, 'kg_embedding_model_name': 'Region', 'region_type': 'sphere', 'embedding_dim': 50, 'normalization_of_entities': 2, 'init_radius': 0, 'reg_lambda': 0.01, 'loss_type': 'MRL', 'margin_loss': 0.5, 'neg_factor': 5, 'corrupt_relations': False, 'filter_negative_triples': False, 'es_metric': 'custom', 'learning_rate': 0.1, 'num_epochs': 1000, 'batch_size': 2048, 'test_batch_size': 4096, 'maximum_number_of_hpo_iters': 5, 'preferred_device': 'gpu', 'pykeen-version': '0.0.26-dev', 'num_entities': 14951, 'num_relations': 1345}

### 